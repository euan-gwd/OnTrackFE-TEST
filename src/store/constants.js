export const GET_BOOKS_BEGIN = 'GET_BOOKS_BEGIN';
export const GET_BOOKS_SUCCESS = 'GET_BOOKS_SUCCESS';
export const GET_BOOKS_ERROR = 'GET_BOOKS_ERROR';
export const CHANGE_PAGE = 'CHANGE_PAGE';
