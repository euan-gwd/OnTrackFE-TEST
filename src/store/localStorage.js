export const loadState = () => {
  try {
    const savedStateInLocalStorage = sessionStorage.getItem('state');
    if (savedStateInLocalStorage === null) {
      return undefined;
    }

    return JSON.parse(savedStateInLocalStorage);
  } catch (error) {
    console.error(error);
    return undefined;
  }
};

export const saveState = state => {
  try {
    const savedState = JSON.stringify(state);
    sessionStorage.setItem('state', savedState);
  } catch (error) {
    console.error(error);
  }
};
