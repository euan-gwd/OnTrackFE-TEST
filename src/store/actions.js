import axios from 'axios';
import * as actionTypes from './constants';

export function getBooks(pageOptions) {
  return (dispatch, getState) => {
    dispatch({ type: actionTypes.GET_BOOKS_BEGIN });
    return axios
      .post(`http://nyx.vima.ekt.gr:3000/api/books/`, pageOptions)
      .then(res => {
        dispatch({
          type: actionTypes.GET_BOOKS_SUCCESS,
          payload: {
            books: res.data.books,
            count: res.data.count,
            page: pageOptions.page,
            itemsPerPage: pageOptions.itemsPerPage
          }
        });
      })
      .catch(error =>
        dispatch({
          type: actionTypes.GET_BOOKS_ERROR,
          payload: {
            error
          }
        })
      );
  };
}
