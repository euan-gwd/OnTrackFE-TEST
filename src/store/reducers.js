import * as actionTypes from './constants';

const initialState = {
  page: 1,
  itemsPerPage: 20,
  isLoading: false,
  error: null,
  books: [],
  totalRecords: 0
};

export function bookReducer(state = initialState, action) {
  switch (action.type) {
    case actionTypes.GET_BOOKS_BEGIN:
      return {
        ...state,
        isLoading: true,
        error: null
      };
    case actionTypes.GET_BOOKS_SUCCESS:
      return {
        ...state,
        books: action.payload.books,
        totalRecords: action.payload.count,
        page: action.payload.page,
        itemsPerPage: action.payload.itemsPerPage,
        isLoading: false,
        error: null
      };
    case actionTypes.GET_BOOKS_ERROR:
      return {
        ...state,
        error: action.payload.error,
        books: [],
        isLoading: false
      };
    default:
      return state;
  }
}
