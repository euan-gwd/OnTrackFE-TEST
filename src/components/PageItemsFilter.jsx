import React from 'react';
import { connect } from 'react-redux';
import ButtonToolbar from 'react-bootstrap/ButtonToolbar';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Button from 'react-bootstrap/Button';
import { getBooks } from '../store/actions';
import { useHistory } from 'react-router-dom';

export const PageItemsFilter = ({ dispatch, totalRecords, page, itemsPerPage }) => {
  let history = useHistory();

  const handleItemsPerPageChange = (newItemsPerPage) => {
    const pageOptions = {
      page: page,
      itemsPerPage: Number(newItemsPerPage)
    };
    dispatch(getBooks(pageOptions));
    history.push(`/${page}`);
  };

  return (
    <ButtonToolbar className="justify-content-end">
      <div></div>
      <ButtonGroup size="sm">
        <Button variant="outline-secondary" disabled>
          Items Per Page
        </Button>
        <Button variant="primary" active={itemsPerPage === 20} onClick={() => handleItemsPerPageChange(20)}>
          20
        </Button>
        <Button variant="primary" active={itemsPerPage === 50} onClick={() => handleItemsPerPageChange(50)}>
          50
        </Button>
        <Button variant="primary" active={itemsPerPage === 75} onClick={() => handleItemsPerPageChange(75)}>
          75
        </Button>
        <Button variant="primary" active={itemsPerPage === 100} onClick={() => handleItemsPerPageChange(100)}>
          100
        </Button>
      </ButtonGroup>
    </ButtonToolbar>
  );
};

const mapStateToProps = (state) => ({
  totalRecords: state.totalRecords,
  page: state.page,
  itemsPerPage: state.itemsPerPage
});

export default connect(mapStateToProps)(PageItemsFilter);
