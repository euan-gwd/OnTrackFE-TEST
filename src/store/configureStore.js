import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { bookReducer } from './reducers';
import { loadState, saveState } from './localStorage';
import thunk from 'redux-thunk';
import throttle from 'lodash/throttle';

const configureStore = () => {
  const persistedState = loadState();

  const store = createStore(bookReducer, persistedState, composeWithDevTools(applyMiddleware(thunk)));

  store.subscribe(
    throttle(() => {
      saveState({
        page: store.getState().page,
        itemsPerPage: store.getState().itemsPerPage,
        books: store.getState().books,
        totalRecords: store.getState().totalRecords
      });
    }, 1000)
  );

  return store;
};

export default configureStore;
