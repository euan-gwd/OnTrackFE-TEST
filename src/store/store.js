import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { bookReducer } from './reducers';
import { loadState } from './localStorage';
import thunk from 'redux-thunk';

const persistedState = loadState();

const store = createStore(
  bookReducer,
  persistedState,
  composeWithDevTools(applyMiddleware(thunk))
);

export default store;
