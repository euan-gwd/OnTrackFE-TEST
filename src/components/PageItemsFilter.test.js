import React from 'react';
import { shallow } from 'enzyme';
import { initialState } from '../store/reducers';
import { PageItemsFilter } from './PageItemsFilter';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3000/'
  }),
  useHistory: () => ({
    pathname: 'localhost:3000/'
  })
}));

describe('PageItemsFilter', () => {
  it('renders as expected', () => {
    const props = {
      ...initialState,
      fetchBooks: jest.fn().mockReturnValue()
    };

    const wrapper = shallow(<PageItemsFilter {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
