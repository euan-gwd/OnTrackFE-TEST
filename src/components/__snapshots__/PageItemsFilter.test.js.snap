// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageItemsFilter renders as expected 1`] = `
<ButtonToolbar
  className="justify-content-end"
  role="toolbar"
>
  <div />
  <ButtonGroup
    role="group"
    size="sm"
    toggle={false}
    vertical={false}
  >
    <Button
      active={false}
      disabled={true}
      type="button"
      variant="outline-secondary"
    >
      Items Per Page
    </Button>
    <Button
      active={true}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      20
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      50
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      75
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="secondary"
    >
      100
    </Button>
  </ButtonGroup>
</ButtonToolbar>
`;
